### ISSUE TITLE
Solution verification with random seeds

### ISSUE THREAD

**Issue by kostmo** [swarm-game/swarm#967]
Some of the more interesting Challenge scenario ideas would entail the player reacting to some "unpredictable" input, either by programming some adaptable algorithmic behavior or utilizing some data structure like a stack.  However, the intention of these scenarios can be easily circumvented by hard-coding the player's robot to the static conditions of the scenario, or, if the scenario utilizes robots with `random` behavior, hard-coding to a particular seed---in other words, "cheesing" the solution.

To account for this, there should exist another "play mode" (perhaps, that is unlocked only after a scenario is completed for the first time): retesting the player's solution with a random seed.  This is analogous to how Leetcode will test a solution against hidden inputs as well as published inputs.

In this mode, the player will be required to submit their solution in advance as a `.sw` file.[^1]  Only then will the seed will be generated (using hardware randomness) and the scenario replayed as a headless simulation.  After completing and evaluating for success or failure, the seed shall be revealed to the player and the player given the option to observe the solution being played out in real time with that seed.  This is the opportunity for the player to discover which edge cases their solution has not addressed.

We will have to decide how many times solutions should be retested in order to award a true "win" of the scenario.  Perhaps there should be a default number (say, 5) that can be overridden per scenario.  It may be, due to the way some scenarios are defined, that a higher number is required to make it likely that the interesting "edge cases" are exercised.

It should be statically determinable whether this higher "win tier" is available for a scenario based on whether any calls to `random` exist in any `program` within the scenario.

In this regime, defining random behavior within certain predictable bounds will be a keystone to scenario design.  The "fairness" (and therefore "enjoyability") of a scenario will be determined by how accurately the goal text describes the behavior of the nondeterministic elements, or how feasible it is to infer the bounds of these elements by observation.

For the purpose of competition (high-scores/records), the stats (completion time) should be tracked for each re-simulation and averaged.  The number of resimulations and the specific values of their seeds shall be observable in the "record".  With a sufficiently large space for seeds (32 bits?), probably no two players will ever have their solutions tested with the same set of seeds, but may duly consider their wins "equivalent" given a sufficient quantity of resimulations and transparency of the seeds.

For complete fairness in a potential tournament-like setting, all players would first submit their solutions, then a set of random seeds would be generated and all players' solutions tested with the same seeds.

[^1]: It may become important to ergonomically "export" solutions that have been authored in the REPL.

**Comment by xsebek** [comment]
You can use system robots for this if it only takes a few ticks to randomise the challenge. 🙂

@byorgey for example in your stack scenario, it should be enough to put base a bit away so it has to `move` for a few ticks and let system robots randomly place 1/0 in a line.

The system support I would like is to not allow player input until the setup has finished (hidden objective). This would also allow us to play intro scenes like flooding the path so now a boat needs to be made.

@kostmo so to answer your question I would stick to testing solution once, though we could run it with more seeds in integration tests.

**Comment by kostmo** [comment]
I had forgotten that by default we do get non-reproducible randomness within the scenario if a particular seed is not specified in the scenario file (and so long as `--seed` is not passed on the command line, which is kind of cheating anyway).  Probably `--seed` should preclude earning high scores, if it's not that way already.

**Comment by kostmo** [comment]
@xsebek I think we still need system support for requiring that the player submits their entire solution before the system robots randomize the playfield.  Otherwise it's still possible to pause the game after randomization and cheese a hardcoded solution.

**Comment by kostmo** [comment]
> @kostmo so to answer your question I would stick to testing solution once,

Just want to make sure I understand your opinion on this---do you mean that we should *retest* once with a random seed?  Or that the original meeting of the win condition is sufficient, with no retesting needed?

**Comment by xsebek** [comment]
IMO no retesting needed, keeping it simple for the players. If they cook up a solution that magically works on a random input then good for them.

That seems to work for Advent of Code, which seems closer in spirit to Swarm then HackerRank which I think used a system like you propose @kostmo.

For our solutions we should of course be much more strict and require it passes at least with a couple of seeds.

**Comment by xsebek** [comment]
Also, as in all things in programming, _we can just simulate it_.

Lets say you want a scenario S, that has to be passed a number of times in a randomised play area.

We can assume its finite area.

Now wall it around, use teleport after finishing it to teleport in a copy of that area that used seed+1.

Repeat as many times as necessary. 🙂 

@kostmo admittedly this is hard to do with current map format, but if we allow map fragments, then it is suddenly much easier.

**Comment by kostmo** [comment]
> we might even have Issue for it (or should have) as we have certainly had that idea before

@xsebek are you thinking of [this discussion](https://github.com/swarm-game/swarm/issues/866#issuecomment-1314869666)?

**Comment by xsebek** [comment]
> Probably --seed should preclude earning high scores, if it's not that way already.

@kostmo do you want to make a separate issue for this?

It seems to me this would solve the most pressing voiced in the description.

If you wanted an entirely fair challenge, then distributing the solutions in a plain text form along with the challenge kind of defeats the purpose. It would be cool if we could make a Web API endpoint to validate a solution, so it could be validated by a headless Swarm running on a server.

**Comment by byorgey** [comment]
I don't think `--seed` should preclude getting high scores, I think they should just be tracked separately.  Set-seed speedrunning is a thing, partly about speed and partly about finding a good seed to run.

**Comment by byorgey** [comment]
Note that this statement from the original issue description:

> It should be statically determinable whether this higher "win tier" is available for a scenario based on whether any calls to `random` exist in any program within the scenario.

is no longer true, now that we have the world construction DSL, since the world can have random elements based on the seed.
